--catalogs option feature

JimmyTournemaine opened the pull request:
Add a new --catalogs option to specify custom catalogs to find schemas prior to the default schemas store.

The initial PR #87 were trying to manage custom schemas when running across multiple files by introducing inline comment in YAML files. As discussed in [this PR](https://github.com/chris48s/v8r/pull/87#issuecomment-877775254). Introducing a `catalogs` option is more accurate to implements a solution for both YAML and JSON files.

Should fix #17 and #61

chris48s commented on src/cli.js:
```diff
@@ -17,11 +17,28 @@ const SCHEMASTORE_CATALOG_URL =
 
-async function getSchemaUrlForFilename(filename, cache, ttl) {
-  const { schemas } = await cachedFetch(SCHEMASTORE_CATALOG_URL, cache, ttl);
+async function getSchemaUrlForFilename(filename, cache, ttl, catalogs) {
+  let schemaLocation;
+  try {
+    const schemas = catalogs.flatMap((c) =>
+      JSON.parse(fs.readFileSync(c, "utf8").toString()).schemas.map((s) => {
+        if (!isUrl(s.url)) {
+          s.url = path.resolve(path.dirname(c), s.url);
+        }
+        return s;
+      })
+    );
+    schemaLocation = getSchemaMatchForFilename(schemas, filename);
+  } catch {
+    const { schemas } = await cachedFetch(SCHEMASTORE_CATALOG_URL, cache, ttl);
+    schemaLocation = getSchemaMatchForFilename(schemas, filename);
+  }
+  return schemaLocation;
+}
```
There are a few problems I can see with this approach.

- I think `--catalogs` should probably accept both a url or a local file (in the same way that `--schema` does). Custom remote catalogs should also be cached
- I don't think we should attempt to merge the catalogs together here. Lets try and match the file in the first schema catalog first, then fall back to the second one, then the third one, etc.. (so there is an order of priority)
- This code (with the try/catch) behaves very strangely (IMO) if we find >1 matches for a given file in a catalog. It treats it the same as zero matches and ploughs on to schemastore. Conflicting `fileMatch` should be a hard error
- Once there are multiple catalogs in play, I think we need to make it clear to the user which catalog the schema we've chosen came from

I think we can probably think of `catalogs` as always being an array which we iterate and implicitly contains `SCHEMASTORE_CATALOG_URL` as the last item in the array (so schemastore is not 'special'). Pulling all that together, I think my suggested implementation of this approach would be something like:

```js
async function getFromUrlOrFile(location, cache, ttl) {
  return isUrl(location)
    ? await cachedFetch(location, cache, ttl)
    : JSON.parse(fs.readFileSync(location, "utf8").toString());
}

async function getSchemaUrlForFilename(catalogs, filename, cache, ttl) {
  for (const [i, catalog] of catalogs.entries()) {
    const { schemas } = await getFromUrlOrFile(catalog, cache, ttl);
    const matches = getSchemaMatchesForFilename(schemas, filename);
    if (matches.length === 1) {
      console.log(`ℹ️ Found schema in ${catalog} ...`);
      return matches[0].url; // Match found. We're done.
    }
    if (matches.length === 0 && i < catalogs.length - 1) {
      continue; // No match found. Try the next catalog in the array.
    }
    if (matches.length > 1) {
      // We found >1 matches in the same catalog. This is always a hard error.
      console.log(
        `Found multiple possible schemas for ${filename}. Possible matches:`
      );
      matches.forEach(function (match) {
        console.log(`${match.description}: ${match.url}`);
      });
    }
    // Either we found >1 matches in the same catalog or we found 0 matches
    // in the last catalog and there are no more catalogs left to try.
    throw new Error(`❌ Could not find a schema to validate ${filename}`);
  }
}

function getSchemaMatchesForFilename(schemas, filename) {
  const matches = [];
  schemas.forEach(function (schema) {
    if ("fileMatch" in schema) {
      if (schema.fileMatch.includes(path.basename(filename))) {
        matches.push(schema);
        return;
      }
      for (const glob of schema.fileMatch) {
        if (minimatch(path.normalize(filename), glob)) {
          matches.push(schema);
          break;
        }
      }
    }
  });
  return matches;
}
```

..and then in `Validator`:

```js
const schemaLocation =
  args.schema ||
  (await getSchemaUrlForFilename(
    [].concat(args.catalogs).concat([SCHEMASTORE_CATALOG_URL]),
    filename,
    cache,
    ttl,
  ));
const schema = await getFromUrlOrFile(schemaLocation, cache, ttl);
```

JimmyTournemaine replied:
Fixed in dad46a

chris48s replied:
yes you're right - `concat()` doesn't mutate :+1:

chris48s commented on src/cli.js:
```diff
@@ -169,6 +192,14 @@ function parseArgs(argv) {
+      default: [],
+      describe:
+        "Local paths of custom catalogs to use prior to schemastore.org",
+    })
```
I can't think of a situation where it would make sense to pass both `--catalogs` and `--schema`. Can you? If not, maybe it would make them mutually exclusive with `.conflicts('schema', 'catalogs')`.?

JimmyTournemaine replied:
Fixed by dad46a136bc280a1863005a941aa9363e03e0a28

Yargs do not handle manage conflictual arguments with default value so I remove this behavior.

chris48s commented on src/cli.js:
```diff
@@ -100,8 +116,15 @@ function Validator() {
       path.extname(filename)
     );
+
+    for (const catalog of args.catalogs) {
+      if (!fs.existsSync(catalog)) {
+        throw new Error(`❌ Catalog not found: ${catalog}`);
+      }
+    }
```
I think we can drop this check and make it 'lazy'. If we follow the "priority" model and the user passed two `--catalogs` and we found a match in the first, it doesn't matter the second was invalid. If we try to read one and it doesn't exist, it will throw anyway. We just throw `ENOENT: no such file or directory` if filename is invalid so (local) catalog can be the same.

JimmyTournemaine replied:
Fixed by dad46a136bc280a1863005a941aa9363e03e0a28

chris48s commented on src/cli.js:
```diff
@@ -35,7 +52,6 @@ async function getSchemaUrlForFilename(filename, cache, ttl) {
   });
-
   if (matches.length === 1) {
     return matches[0].url;
```
I have been pretty slack so far about assuming the structure of the schemastore catalog, but once the catalog can be arbitrary user-supplied JSON I think we need to be more defensive about dealing with the possibility that the schema catalog might not be valid. One option is that (of course :) ) there is a JSON schema for validating schema catalogs https://json.schemastore.org/schema-catalog.json so we could check the supplied catalog against that before we try to use it, although the downside of that is it requires a bunch of fields we're not actually going to use. We should a least consider the possibility that the catalog might not match our assumptions (contains a key called `schemas` which is an array[] of objects{} which must have a `.url` property. There is also another reason why maybe we don't want to strictly validate that a user-supplied catalog matches https://json.schemastore.org/schema-catalog.json which we will come onto in the next comment..

JimmyTournemaine replied:
Fixed by dad46a136bc280a1863005a941aa9363e03e0a28

chris48s commented on testfiles/catalog-local.json:
```diff
@@ -0,0 +1,13 @@
+        "valid.json",
+        "valid.yml"
+      ],
+      "url": "./schema.json"
```
So I hadn't really thought about this until I saw this code, but this opens a bit of a can of worms.. I'll try and summarise my thoughts.

- This `url` is not a URL - its a local path, which is confusing
- The schema catalog schema says a schema URL is

  ```json
  "url": {
    "description": "An absolute URL to the schema location",
    "type": "string",
    "format": "uri",
    "pattern": "^https://"
  }
  ```
  so this is now not strictly a schema catalog.
- If we ignore this and allow `url` to be a path to a local file this introduces an important question: if the path to the schema is relative, what is it relative to?
  I think in your original implementation you're saying the path of the schema is relative to the path of the catalog file?
  In the one I've suggested above the path is assumed to be relative to the execution context (i.e: we just treat it the same as an arg passed to `--schema`), so this would become `./testfiles/schema.json`
  That's kind of wierd, but the concept of the schema path being relative to the catlog path gets even wierder if the catalog is a remote URL.

I'm undecided on where to go with this.. I think there are two options worth considering, neither of which are completely satisfying:
- Only allow `schema.url` be be an actual URL.
- Break with the schema catalog definition, allow `schema.url` to also be a path (absolute or relative). If the path is relative interpret it relative to the execution context.

The first is more technically pure. A user-supplied schema catalog has the same rules as schemastore. However it means we can specify either a remote or local schema with `--schema` but there is no way to specify we want to use a local schema with `--catalogs`
The second is annoying because `schema.url` might be.. not a URL. We're inventing our own thing rather than leveraging an existing format and your catalog might contain relative paths that don't make sense relative to the catalog itself, depending on where you put it. It does give us a way to address the local schema use-case though. I'm somewhat torn..

Interested in your thoughts on the tradeoffs here as someone who will presumably actually use this feature...

JimmyTournemaine replied:
For now, I have implemented a solution that manage local schemas relative to the execution context despite the expected catalog format in dad46a136bc280a1863005a941aa9363e03e0a28.

It also allow a strange behavior (but that could make sense to someone ??). Take a look at the "should return 0 when file is valid (with auto-detected schema from custom remote catalog matching relativly defined schema)" test.

chris48s replied:
> This I why I remove the $schema property in test files.

Ah yes. I had not spotted this.

> I was first thinking about adding a `path` property to match against local schemas

Yes this is another option I considered. Again, it is not completely satisfying but on reflection maybe better then abusing `url` if we need to diverge anyway.

> I do not have a need for using local schemas in a local catalog

OK. This is a useful data point.
I know this is becoming a bit annoying to work on as I am changing the requirements as we go along, but I think a "safe" starting point for this feature is we say `schema.url` can only be a URL. This is unambiguously correct wherever the catalog lives. It makes it easy to understand/explain the feature and it solves the problem you (and some other users) are having. Doing this also doesn't prohibit us (me) from adding the ability to use a local schema (in some way or another - implementation to be decided) at a later date.

Essentially what worries me is that all the solutions we've come up with so far between the two of us for specifing local schemas are a bit dubious in some way - we haven't really hit on anything that is obviously correct or intuitive. What concerns me as a maintainer is that I merge a solution which is somewhat half-baked, people start relying on it and then it becomes hard to change. Of course, the project is pre-1.0 so I can make breaking changes at will, but I would rather not do this :) It may also be easier to understand what the right solution should be once we have real examples of users who want to create a catalog referencing local schemas (or maybe they never materialize and it becomes a non-issue).

Lets start with:

- catalog can be a remote URL or a local path
- `schema.url` is always a URL

We can get the feature landed on that basis and worry about extending it as another issue.

chris48s reviewed (commented):
Thanks for having a look at this. I have had a look and left some comments. It is useful to have a look at an implementation although in doing so I realise there are still some questions to answer.

I haven't really read over the tests properly yet. Lets work out exactly what the feature behaviour looks like first and then I'll review the tests, but I will quickly note here if we want `--catalogs` to be an array arg, we should have at least one test verifying the correct order of processing if >1 custom catalogs are specified including tests for the situation where a match pattern exists in >1 catalogs.

JimmyTournemaine reviewed (commented):
Thanks for your review, I'will take into account all you remarks and add some unit tests for the cases I was not thinking about.

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -84,7 +87,9 @@ describe("CLI", function () {
+      return cli({
+        filename: "./testfiles/valid.json",
+      }).then((result) => {
         assert.equal(0, result);
```
Just for information, chai assert is expected: `equal(actual, expected, [message])` but all the tests reverse the `expected` and `actual` order, generating ununderstandable messages on test failures.
I still kept the same order for the new tests.

chris48s replied:
Good catch. I'll raise another issue to deal with this. Lets not try and boil the whole ocean in one pull request.
Also, thanks for keeping with the existing order. It will be easier to switch them all after we get this merged if they're all following the same convention to start with (even if that is the wrong way round).
I seem to be accumulating a variety of new exciting issues to solve..

JimmyTournemaine commented on src/test-helpers.js:
```diff
@@ -16,6 +17,7 @@ function setUp(messages) {
 }
 
 function tearDown() {
+  cache.resetCallCounter();
```
Prevent falsy cyclic detection

chris48s commented on src/cache.js:
```diff
@@ -63,4 +63,8 @@ async function cachedFetch(url, cache, ttl) {
 }
 
-module.exports = { cachedFetch, expire };
+function resetCallCounter() {
+  callCounter = {};
+}
```
I think the proper solution to this will probably be to encapsulate this in a class so every test can instantiate a new cache object (with a clean `callCounter`) but I agree we can work around it for now and punt it to an issue for another day.

chris48s commented:
I haven't done another full review of the diff but I have replied to a bunch of comments above. I think there is now a clear path to getting this done :crossed_fingers:

JimmyTournemaine commented:
Now catalogs are validated using the catalog JSON validation from schema store.
I replace the *custom checks* by a the real schema validation, so I moved the `resolver` declaration in the `validate` function to reuse it and add a control on the `schemas` field (that is not defined as required by the catalog schema).
I updated the unit tests.

Let me know if you see anything to change.

chris48s commented:
Thanks the updates :+1: I've had a really quick skim over the diff and things look broadly sensible but I will make time over the next few days to properly read/test it.

chris48s commented on src/cli.js:
```diff
@@ -15,11 +15,57 @@ const logging = require("./logging.js");
   "https://www.schemastore.org/api/json/catalog.json";
+
+const SCHEMASTORE_CATALOG_SCHEMA_URL =
+  "https://www.schemastore.org/api/json/schema-catalog.json";
```
```suggestion
  "https://json.schemastore.org/schema-catalog.json";
```

This should be https://json.schemastore.org/schema-catalog.json
https://www.schemastore.org/api/json/schema-catalog.json is an error

JimmyTournemaine replied:
You're right. I was making some tests and I forget to undo it before committing. Resolved by 7b3428af9bfd65407dfceebcb7e6bd12de6de600.

chris48s commented on src/cli.js:
```diff
@@ -15,11 +15,57 @@ const logging = require("./logging.js");
+    const { schemas } = catalog;
+    const matches = getSchemaMatchesForFilename(schemas, filename);
+    if (matches.length === 1) {
+      console.log(`ℹ️ Found schema in ${catalog} ...`);
```
```suggestion
      console.log(`ℹ️ Found schema in ${catalogLocation} ...`);
```

With `catalog` This prints `Found schema in [object Object]`

JimmyTournemaine replied:
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

chris48s commented on src/cli.spec.js:
```diff
@@ -9,11 +9,39 @@ const { setUp, tearDown } = require("./test-helpers.js");
 
 describe("CLI", function () {
+  // Moch the catalog validation schema
+  beforeEach(() => {
```
Not specifically related to this PR. I should have already done it, but there is a lot of network IO going on that we need to mock. Lets add `nock.disableNetConnect()` to the start of `beforeEach()` too. This will stop nock from letting through any request to the network so if we forget to mock a request under test it will fail.

JimmyTournemaine replied:
I did not know this feature, I add it in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

chris48s commented on src/cli.spec.js:
```diff
@@ -9,11 +9,39 @@ const { setUp, tearDown } = require("./test-helpers.js");
 chai.use(require("chai-as-promised"));
 
 describe("CLI", function () {
+  // Moch the catalog validation schema
```
```suggestion
  // Mock the catalog validation schema
```

JimmyTournemaine replied:
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

chris48s commented on src/cli.spec.js:
```diff
@@ -300,6 +456,81 @@ describe("CLI", function () {
+        assert.equal(1, result);
+        mock.done();
+        expect(messages.error).to.contain(
+          "❌ Malformed catalog at https://example.com/catalog.json",
```
```suggestion
          "❌ Malformed catalog at https://example.com/catalog.json"
```

appease the lint gods :pray:

JimmyTournemaine replied:
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

chris48s commented on src/cli.js:
```diff
@@ -169,6 +206,14 @@ function parseArgs(argv) {
+      alias: "c",
+      array: true,
+      describe:
+        "Local paths of custom catalogs to use prior to schemastore.org",
```
This can also be a URL

JimmyTournemaine replied:
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

chris48s reviewed (commented):
I found a bit of time to have a look over this and do some local testing.
I still need to properly read through the test cases. I will get to this tomorrow or at the weekend.

chris48s commented on src/cli.spec.js:
```diff
@@ -300,6 +456,81 @@ describe("CLI", function () {
+          "❌ Malformed catalog at ./testfiles/catalog-local.json"
+        );
+      });
+    });
```
Nice :+1:  This block of tests for the various failure/invalid cases are very comprehensive :)

chris48s reviewed (commented):
I've added some more comments. I think it is all minor stuff left now though. In many cases I've proposed an edit you can just click to accept.
I have a bit of time free next week when I can work on some of the follow-up jobs like reversing the test assertions and refactoring the cache module so it would be amazing if we can get this PR finished and merged in the next few days.

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -117,6 +149,130 @@ describe("CLI", function () {
       });
     });
 
+    it("should return 0 when file is valid (with auto-detected schema from custom local catalog matching remote schema)", function () {
```
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -117,6 +149,130 @@ describe("CLI", function () {
+      });
+    });
+
+    it("should return 0 when file is valid (with auto-detected schema from custom remote catalog matching remote schema)", function () {
```
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -117,6 +149,130 @@ describe("CLI", function () {
+      });
+    });
+
+    it("should return 0 when file is valid (with auto-detected schema from custom catalog fallbacking to the next catalog)", function () {
```
Sorry for my bad English ^^
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -117,6 +149,130 @@ describe("CLI", function () {
+      });
+    });
+
+    it("should return 0 when file is valid (with auto-detected schema from custom catalog fallbacking to schemastore.org)", function () {
```
Fixed in 7b3428af9bfd65407dfceebcb7e6bd12de6de600

JimmyTournemaine commented on src/cli.spec.js:
```diff
@@ -117,6 +149,130 @@ describe("CLI", function () {
+      });
+    });
+
+    it("should return 99 when file is invalid (with auto-detected schema and custom catalogs)", function () {
```
You're right, it was a useful test during the development phase to detect when I was breaking things.
The test has no value now, I removed it in fca27aab2c13f6fce061142fa0b4cca4416af203.

chris48s reviewed (approved):
Nice - thanks for your work on this :+1: 
I'll push a new release to NPM tomorrow
